File: backend/app/schemas/alter_ego.py
"""
Alter Ego API schemas
"""
from typing import List, Literal, Optional

from pydantic import BaseModel, Field

from app.models.alter_ego_dna import AlterEgoDnaJson, DnaStatus, PrimaryLanguage


class ExtractRequest(BaseModel):
    exemplars: List[str] = Field(..., min_length=1, max_length=3)
    language: PrimaryLanguage = "zh-TW"


class ExtractResponse(BaseModel):
    dna_json: AlterEgoDnaJson
    dna_version_id: str
    dna_status: Literal["active"] = "active"


class ComposeRequest(BaseModel):
    """New public composer. Does not change PreviewRequest.platform."""

    platform: Literal["facebook", "instagram", "threads"]
    style: Literal[
        "professional", "casual", "humorous", "storytelling", "educational"
    ]
    max_chars: int = Field(..., ge=50, le=5000)
    part: Literal["all", "title", "body", "hashtags"] = "all"
    language: PrimaryLanguage = "zh-TW"
    topic_id: Optional[str] = Field(default=None, max_length=64)
    topic_title: str = Field(default="", max_length=300)
    context_summary: str = Field(default="", max_length=1500)


class ComposeResponse(BaseModel):
    titles: List[str]
    body: str
    hashtag_sets: List[List[str]]
    credits_charged: int = 1
    balance_after: int = 0
    max_chars: int = 150


class PreviewRequest(BaseModel):
    platform: Literal["facebook", "threads", "x"]
    topic_hint: str = Field(default="", max_length=200)
    language: Optional[PrimaryLanguage] = Field(
        default=None,
        description="UI 語言覆寫（優先於 DNA language_primary）",
    )
    context_summary: Optional[str] = Field(
        default="",
        max_length=1500,
        description="主題事實摘要 (summary_flash)，作為防偏題錨定",
    )
    base_content: Optional[str] = Field(
        default="",
        max_length=4000,
        description="已生成之基礎短文，供平台改寫",
    )


class PreviewResponse(BaseModel):
    platform: str
    preview_text: str
    soul_text: str = ""
    shell_constraints: str


class DnaStatusResponse(BaseModel):
    dna_status: DnaStatus
    current_dna_version_id: Optional[str] = None
    has_dna: bool = False


class RollbackRequest(BaseModel):
    snapshot_id: str = Field(..., min_length=8, max_length=64)


class SkipResponse(BaseModel):
    dna_status: Literal["skipped"] = "skipped"


class AdoptCopyRequest(BaseModel):
    platform: Literal["facebook", "threads", "x"]
    topic_id: Optional[str] = Field(default=None, max_length=64)
    preview_text: str = Field(..., min_length=1, max_length=8000)


class AdoptCopyResponse(BaseModel):
    logged: bool = True
    event: Literal["adopted_without_edit"] = "adopted_without_edit"


class FeedbackRequest(BaseModel):
    action: Literal["like", "dislike"]
    topic_id: Optional[str] = Field(default=None, max_length=64)
    comment: Optional[str] = Field(default=None, max_length=500)


class FeedbackResponse(BaseModel):
    logged: bool = True

File: frontend/src/api/alterEgo.ts
/**
 * Alter Ego API — onboarding / DNA（AE-1d）
 */
import { fetchAPI, fetchAPIEnvelope } from './client';
import { normalizeUiLanguage } from '@/lib/topicLanguages';

export type DnaStatus = 'pending' | 'active' | 'skipped' | 'legacy_only';
export type AlterEgoPlatform = 'facebook' | 'threads' | 'x';
export type ComposePlatform = 'facebook' | 'instagram' | 'threads';
export type ComposeStyle =
  | 'professional'
  | 'casual'
  | 'humorous'
  | 'storytelling'
  | 'educational';
export type ComposePart = 'all' | 'title' | 'body' | 'hashtags';

export interface DnaStatusResponse {
  dna_status: DnaStatus;
  current_dna_version_id?: string | null;
  has_dna: boolean;
}

export interface ExtractResponse {
  dna_json: Record<string, unknown>;
  dna_version_id: string;
  dna_status: 'active';
}

export interface PreviewResponse {
  platform: string;
  preview_text: string;
  soul_text: string;
  shell_constraints: string;
}

export interface ComposeResponse {
  titles: string[];
  body: string;
  hashtag_sets: string[][];
  credits_charged: number;
  balance_after: number;
  max_chars: number;
}

export const alterEgoApi = {
  /** 勿走 unwrap data，避免誤拆壞 dna_status */
  getStatus: () => fetchAPIEnvelope<DnaStatusResponse>('/alter-ego/status'),

  skip: () => fetchAPI<{ dna_status: 'skipped' }>('/alter-ego/skip', { method: 'POST' }),

  /** extract／preview 各含 1～2 次 DeepSeek；預設 10s 會誤殺（Request timeout 10000ms） */
  extract: (exemplars: string[], language: string = 'zh-TW') =>
    fetchAPI<ExtractResponse>('/alter-ego/extract', {
      method: 'POST',
      body: JSON.stringify({ exemplars, language }),
      timeout: 120000,
    }),

  preview: (
    platform: AlterEgoPlatform,
    topicHint: string = '',
    language?: string,
    contextSummary?: string,
    baseContent?: string
  ) =>
    fetchAPI<PreviewResponse>('/alter-ego/preview', {
      method: 'POST',
      body: JSON.stringify({
        platform,
        topic_hint: topicHint,
        ...(language ? { language: normalizeUiLanguage(language) } : {}),
        ...(contextSummary ? { context_summary: contextSummary } : {}),
        ...(baseContent ? { base_content: baseContent } : {}),
      }),
      timeout: 120000,
    }),

  rollback: (snapshotId: string) =>
    fetchAPI<ExtractResponse>('/alter-ego/dna/rollback', {
      method: 'POST',
      body: JSON.stringify({ snapshot_id: snapshotId }),
    }),

  adoptCopy: (payload: { platform: AlterEgoPlatform; topic_id?: string; preview_text: string }) =>
    fetchAPI<{ logged: boolean; event: string }>('/alter-ego/adopt-copy', {
      method: 'POST',
      body: JSON.stringify({
        platform: payload.platform,
        topic_id: payload.topic_id,
        preview_text: payload.preview_text,
      }),
    }),

  compose: (payload: {
    platform: ComposePlatform;
    style: ComposeStyle;
    max_chars: number;
    part?: ComposePart;
    language: string;
    topic_id?: string;
    topic_title?: string;
    context_summary?: string;
  }) =>
    fetchAPI<ComposeResponse>('/alter-ego/compose', {
      method: 'POST',
      body: JSON.stringify({
        platform: payload.platform,
        style: payload.style,
        max_chars: payload.max_chars,
        part: payload.part || 'all',
        language: normalizeUiLanguage(payload.language),
        topic_id: payload.topic_id,
        topic_title: payload.topic_title || '',
        context_summary: payload.context_summary || '',
      }),
      timeout: 120000,
    }),
};
